apps/emqx/src/emqx_channel.erl

```diff
@@ -398,8 +398,10 @@ handle_in(?PACKET(_), Channel = #channel{conn_state = ConnState}) when
     handle_out(disconnect, ?RC_PROTOCOL_ERROR, Channel);
 handle_in(Packet = ?PUBLISH_PACKET(_QoS), Channel) ->
     case emqx_packet:check(Packet) of
-        ok -> process_publish(Packet, Channel);
-        {error, ReasonCode} -> handle_out(disconnect, ReasonCode, Channel)
+        ok ->
+            emqx_external_trace:trace_process_publish(Packet, Channel, fun process_publish/2);
```

Comment:
should this work?
```suggestion
            emqx_external_trace:trace_process_publish(Packet, fun(PacketWithTrace) -> process_publish(PacketWithTrace) end);
```

Reply:
ok, need to extrace client id later.
channel_attributes does not seem to be an expensive operation, so we can extract it and pass it down as arg